### DISCUSSION TITLE
Global LoadingStrategy.JOINED with embeddables generates "Cannot read properties of undefined (reading 'forEach')"

### DISCUSSION THREAD

**Discussion by robbyemmert** [mikro-orm/mikro-orm#4557] [General]
**Describe the bug**
When setting loadStrategy: LoadStrategy.JOINED globally, querying findOne on entities with Embedded entities (object:false) produces the following error: Cannot read properties of undefined (reading 'forEach')

**Stack trace**
```
"DriverException: Cannot read properties of undefined (reading 'forEach')",
        "    at PostgreSqlExceptionConverter.convertException (/my/project/path/node_modules/@mikro-orm/core/platforms/ExceptionConverter.js:8:16)",
        "    at PostgreSqlExceptionConverter.convertException (/my/project/path/node_modules/@mikro-orm/postgresql/PostgreSqlExceptionConverter.js:42:22)",
        "    at PostgreSqlDriver.convertException (/my/project/path/node_modules/@mikro-orm/core/drivers/DatabaseDriver.js:197:54)",
        "    at /my/project/path/node_modules/@mikro-orm/core/drivers/DatabaseDriver.js:201:24",
        "    at processTicksAndRejections (node:internal/process/task_queues:96:5)",
        "    at async PostgreSqlDriver.find (/my/project/path/node_modules/@mikro-orm/knex/AbstractSqlDriver.js:49:24)",
        "    at async PostgreSqlDriver.findOne (/my/project/path/node_modules/@mikro-orm/knex/AbstractSqlDriver.js:67:21)",
        "    at async SqlEntityManager.findOne (/my/project/path/node_modules/@mikro-orm/core/EntityManager.js:327:22)",
        "    at async JobService.getOne (/my/project/path/dist/utils/data/data-service.js:45:20)",
        "    at async JobController.getJob (/my/project/path/dist/modules/jobs/controllers/job-controller.js:30:18)",
        "",
        "previous TypeError: Cannot read properties of undefined (reading 'forEach')",
        "    at /my/project/path/node_modules/@mikro-orm/knex/query/QueryBuilderHelper.js:172:30",
        "    at Array.forEach (<anonymous>)",
        "    at QueryBuilderHelper.processJoins (/my/project/path/node_modules/@mikro-orm/knex/query/QueryBuilderHelper.js:163:30)",
        "    at QueryBuilder.getQueryBase (/my/project/path/node_modules/@mikro-orm/knex/query/QueryBuilder.js:751:29)",
        "    at QueryBuilder.getKnexQuery (/my/project/path/node_modules/@mikro-orm/knex/query/QueryBuilder.js:364:25)",
        "    at QueryBuilder.execute (/my/project/path/node_modules/@mikro-orm/knex/query/QueryBuilder.js:439:28)",
        "    at PostgreSqlDriver.find (/my/project/path/node_modules/@mikro-orm/knex/AbstractSqlDriver.js:49:46)",
        "    at PostgreSqlDriver.findOne (/my/project/path/node_modules/@mikro-orm/knex/AbstractSqlDriver.js:67:32)",
        "    at SqlEntityManager.findOne (/my/project/path/node_modules/@mikro-orm/core/EntityManager.js:327:38)",
        "    at processTicksAndRejections (node:internal/process/task_queues:96:5)"
```

**To Reproduce**
1. Create an entity with an embedded entity as a property
2. Set your global loadStrategy to be LoadStrategy.JOINED
3. Query findOne on your base entity

**Expected behavior**
Embedded entities should be queryable with any loadStrategy since they aren't technically a separate entity.

**Additional context**
I traced the issue to function processJoins in QueryBuilderHelper.js. The function is looping over joins on the queried entity. The erroring join is the embedded entity, since it's trying to forEach off of the embedded entity's primary keys. Not sure why it thinks the embedded entity is a join. Technically it's not a relation.

Here is my schema:
My embeddable
```ts
@Embeddable()
export class BuildingDetails {
// ...
}
```

Embedded attribute:
```ts
@Embedded(() => BuildingDetails, {
    nullable: true,
    prefix: false,
  })
  buildingDetails = new BuildingDetails();
  
```

**Versions**

| Dependency | Version |
| - | - |
| node | v16.18.0 |
| typescript | ^4.1.3 |
| mikro-orm | ^5.7.13 |
| your-driver | postgresql |

**Comment by B4nan** [comment]
I need to see complete reproduction for such issues, words are simply not enough. 

> Embedded entities should be queryable with any loadStrategy since they aren't technically a separate entity.

We have tests for this use case, it works fine in there.

> I traced the issue to function processJoins in QueryBuilderHelper.js. The function is looping over joins on the queried entity. The erroring join is the embedded entity, since it's trying to forEach off of the embedded entity's primary keys. Not sure why it thinks the embedded entity is a join. Technically it's not a relation.

Having complete executable repro is the first step, not debugging and suggesting things :] I can't fix anything unless I have a failing test for it. Your repro is not even mentioning any relations, yet you say that you traced it to be related to joining... See where I am heading? :]

**Comment by bakura10** [comment]
I am also facing the same issue (on v6 beta). Assuming an entity "shop" with an embeddable "brand" that contains a ManyToOne reference to image, this works:

```
await walletItemsRepository.findAll({populate: ['shop.brand.logoImage']});
```

But this one throws the OP's error:

```
await walletItemsRepository.findAll({populate: ['shop.brand.logoImage'], strategy: 'joined'});
```

I will try to setup a test case later next week in case you are not able to reproduce it.
